lib/docassemble/docassemble_api_interface.js

```diff
@@ -0,0 +1,377 @@
+      + `projects with the base name ${ session_vars.get_project_base_name() }. `
+      + `File an issue at https://github.com/SuffolkLITLab/ALKiln/issues.`;
+    throw too_many_project_names_msg;
+  }
+};  // Ends da_i.loop_to_create_project()
+
+
+// node -e 'require("./lib/github_integration/docassemble_api_interface.js").pull()'
+da_i.pull = async function ( pull_options, server_wait_options ) {
+  /** Pull github code into an existing Project. */
+
+  log.info({ pre: `Trying to pull code from the "${ session_vars.get_branch_name() }" branch.` });
+
+  // Default timeouts
+  let { timeout } = pull_options ||  { timeout: 75 * 1000 };
+
+  try {
+
+    let response = await _da.pull( timeout );
+
+    // If server didn't need to restart
+    if ( response && response.status === 204 ) {
+      log.debug({ pre: `Status 204 on pull. Did not need the docassemble server to restart.` });
+      return null;
+    }
+
+    // If server needed to restart because of the pulled code
+    if ( response && response.status === 200 ) {
+      log.debug({ pre: `Pull task id: ${ response.data.task_id }` });
+      return response.data.task_id;
+    }
+
+  } catch (error) {
+    // Pulling should not error
+    log.info({ pre: `Pulling to the docassemble Project ran into an unexpected error. See https://docassemble.org/docs/api.html#playground_pull` });
+    throw error;
+  }
+};  // Ends da_i.pull()
+
+
+// node -e 'require("./lib/github_integration/docassemble_api_interface.js").has_task_finished()'
+da_i.has_task_finished = async function ( task_id, options ) {
+  /** Check task status if it exists. If not, return true as task did
+  *    not require a server restart. See da_i.pull().
+  *    See https://docassemble.org/docs/api.html#restart_status. */
+
+  // Defaults
+  let { timeout } = options ||  { timeout: 3 * 1000 };
+
+  try {
+
+    // If server did need to restart, check if it's done.
+    let response = await _da.has_task_finished( task_id, timeout );
+    log.debug({ pre: `Task status: ${ response.status }. Docassemble server status: ${ response.data.status }. Task id: ${ task_id }` });
+    
+    if ( response.data.status === 'completed' ) {
+      log.info({ pre: `The docassmble server restarted successfully.` });
+      return true;
+    
+    } else if ( response.data.status === 'working' ) {
+      log.info({ pre: `The docassmble server is still working on it.` });
+      return false;
+
+    } else if ( response.data.status === 'unknown' ) {
+      log.info({
+        pre: `\`task_id\` "${ task_id }" has expired. If it's been `
+          + `less than 1 hour since the task started, the task is `
+          + `complete. See https://docassemble.org/docs/api.html#restart_status.`,
+        data: response,
+      });
+
+      // Server is working, but status unexpected? What to return? Throw?
+      // We shouldn't even get here according to the API docs, but it feels
+      // weird to have no default behavior
```

Comment:
We have decided that the meaning of this comment is clearer if it is moved above the log.info(...) line.